apps/ui/Application.cpp

```diff
@@ -0,0 +1,385 @@
+        for (const auto& line : m_brayns.getKeyboardHandler().help())
+            ImGui::Text("%s", line.c_str());
+        ImGui::End();
+    }
+}
+
+void Application::guiRender()
+{
+    ImGui::Render();
+    glViewport(0, 0, m_width, m_height);
+    ImGui_ImplOpenGL3_RenderDrawData(ImGui::GetDrawData());
+}
+
+void Application::guiShutdown()
+{
+    ImGui_ImplOpenGL3_Shutdown();
+    ImGui_ImplGlfw_Shutdown();
+    ImGui::DestroyContext();
+}
+
+void Application::initOpenGL()
+{
+    glGenTextures(1, &m_fbTexture);
+    glEnable(GL_TEXTURE_2D);
+    glBindTexture(GL_TEXTURE_2D, m_fbTexture);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
+    glBindTexture(GL_TEXTURE_2D, 0);
+}
+
+void Application::render()
+{
+    glViewport(0, 0, m_width, m_height);
+    glClearColor(0.f, 0.f, 0.f, 1.f);
+    glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
+
+    m_brayns.commitAndRender();
+
+    size_t offset = 0;
+    for (auto frameBuffer : m_brayns.getEngine().getFrameBuffers())
+    {
+        GLenum format = GL_RGBA;
+        switch (frameBuffer->getFrameBufferFormat())
+        {
+        case brayns::FrameBufferFormat::bgra_i8:
+            format = GL_BGRA;
+            break;
+        case brayns::FrameBufferFormat::rgb_i8:
+            format = GL_RGB;
+            break;
+        default:
+            format = GL_RGBA;
+        }
+
+        const auto& frameSize = frameBuffer->getFrameSize();
+
+        frameBuffer->map();
+        glMatrixMode(GL_PROJECTION);
+        glLoadIdentity();
+        glOrtho(0, frameSize.x, 0.0f, frameSize.y, -1.0f, 1.0f);
+
+        glMatrixMode(GL_MODELVIEW);
+        glLoadIdentity();
+        glViewport(offset, 0, frameSize.x, frameSize.y);
+
+        GLenum type = GL_FLOAT;
+        const GLvoid* buffer = 0;
+        switch (m_frameBufferMode)
+        {
+        case FrameBufferMode::COLOR:
+            type = GL_UNSIGNED_BYTE;
+            buffer = frameBuffer->getColorBuffer();
+            break;
+        case FrameBufferMode::DEPTH:
+            format = GL_LUMINANCE;
+            buffer = frameBuffer->getDepthBuffer();
+            break;
+        default:
+            glClearColor(0.f, 0.f, 0.f, 1.f);
+            glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
+        }
+
+        if (buffer)
+        {
+            glBindTexture(GL_TEXTURE_2D, m_fbTexture);
+            glTexImage2D(GL_TEXTURE_2D, 0, format, frameBuffer->getSize().x,
+                         frameBuffer->getSize().y, 0, format, type, buffer);
+
+            glBegin(GL_QUADS);
+            glTexCoord2f(0.f, 0.f);
+            glVertex3f(0, 0, 0);
+            glTexCoord2f(0.f, 1.f);
+            glVertex3f(0, frameSize.y, 0);
+            glTexCoord2f(1.f, 1.f);
+            glVertex3f(frameSize.x, frameSize.y, 0);
+            glTexCoord2f(1.f, 0.f);
+            glVertex3f(frameSize.x, 0, 0);
+            glEnd();
```

Comment:
Since you're modernizing this code, you could try replacing the immediate mode draw calls with VBOs.